compiler/rustc_typeck/src/check/cast.rs

```diff
@@ -973,6 +984,50 @@ impl<'a, 'tcx> CastCheck<'tcx> {
             }
         }
     }
+
+    fn lossy_provenance_ptr2int_lint(&self, fcx: &FnCtxt<'a, 'tcx>, t_from: CastTy<'tcx>) {
+        fcx.tcx.struct_span_lint_hir(
+            lint::builtin::LOSSY_PROVENANCE_CASTS,
+            self.expr.hir_id,
+            self.span,
+            |err| {
+                let mut err = err.build(&format!(
+                    "under strict provenance it is considered bad style to cast pointer `{}` to integer `{}`",
+                    self.expr_ty, self.cast_ty
+                ));
+
+                if let CastTy::FnPtr = t_from {
+                    err.help(
+                        "use `(... as *const u8).addr()` to obtain \
+                         the address of a function pointer",
+                    );
+                } else {
+                    err.help("use `.addr()` to obtain the address of a pointer");
+                }
```

Comment:
These should probably be made into [suggestions](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic/struct.Diagnostic.html#method.span_suggestion)? That means that the errors will contain a "preview" of the suggested code, and tools like `rustfix` can apply suggestions automatically.

I'm guessing the ptr -> int would always be correct (so [`MachineApplicable`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/enum.Applicability.html#variant.MachineApplicable)) and the int -> ptr would have placeholders (so [`HasPlaceholder`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/enum.Applicability.html#variant.HasPlaceholders))?

Reply:
Ah, I didn't know. Although I am a bit hesitant to make the ptr -> int lint machine applicable since there is recent discussion in #95588 of whether `.addr()` and `as usize` should indeed be identical or not.